### PR TITLE
SD-521: fetch erc20 token balances for specified block number

### PR DESCRIPTION
**Pull request by domagojlatecki**

### CODE REVIEW

File: src/main/kotlin/com/ampnet/payoutservice/blockchain/BlockchainServiceImpl.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                throw InternalException(
+                    ErrorCode.BLOCKCHAIN_CONTRACT_EVENT_READ_ERROR,
+                    "Error processing contract transfer event"
+                )
+            }
+        )
+
+        logger.debug { "Found ${accounts.size} holder addresses for ERC20 contract: $erc20ContractAddress" }
+
+        contract.setDefaultBlockParameter(endBlockParameter)
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/blockchain/BlockchainServiceImpl.kt:53]: Why is this needed? To check the balance at the end block?

File: src/main/kotlin/com/ampnet/payoutservice/config/ApplicationProperties.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.ampnet.payoutservice.config
+
+import org.springframework.boot.context.properties.ConfigurationProperties
+import org.springframework.context.annotation.Configuration
+
+@Configuration
+@ConfigurationProperties(prefix = "com.ampnet.payoutservice")
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/config/ApplicationProperties.kt]: Maybe we could start removing `com.ampnet` and use just `payoutservice` because it is shorter and we don't lose any information. What do you think?

File: src/integTest/kotlin/com/ampnet/payoutservice/blockchain/BlockchainServiceIntegTest.kt [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+        contract.applyWeb3jFilterFix(startBlock, endBlock1)
+
+        suppose("Some additional transactions of ERC20 token are made") {
+            contract.transferAndMine(accounts[1].address, BigInteger("900"))
+            contract.transferAndMine(accounts[5].address, BigInteger("1000"))
+            contract.transferAndMine(accounts[6].address, BigInteger("2000"))
+        }
+
+        val endBlock2 = HardhatTestContainer.blockNumber()
```
Comment (githinho) [inline src/integTest/kotlin/com/ampnet/payoutservice/blockchain/BlockchainServiceIntegTest.kt]: It would be clearer to read if `endBlock2` is placed above or inside the second `verify`

File: src/main/kotlin/com/ampnet/payoutservice/exception/InternalException.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.ampnet.payoutservice.exception
+
+class InternalException(val errorCode: ErrorCode, exceptionMessage: String) : Exception(exceptionMessage) {
+    companion object {
+        private const val serialVersionUID: Long = -2656013978175834059L
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/payoutservice/exception/InternalException.kt:5]: What is the purpose of this value?

File: src/main/kotlin/com/ampnet/payoutservice/blockchain/BlockchainServiceImpl.kt (hunk 1 above)
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/payoutservice/blockchain/BlockchainServiceImpl.kt:53]: This is used to tell web3j to fetch token balances for specified block number. We are building a history from `startBlock` (which should either be the chain's genesis block or the block at which token was deployed, this is used for efficiency) to `endBlock` (which is the block at which the payout is being made). So it is important that me know the balances at `endBlock` for all accounts as that is the block number for which the payout is being made.

File: src/main/kotlin/com/ampnet/payoutservice/config/ApplicationProperties.kt (hunk 2 above)
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/payoutservice/config/ApplicationProperties.kt]: Yeah, we could start doing that. I will remove the prefix.

File: src/main/kotlin/com/ampnet/payoutservice/exception/InternalException.kt (hunk 4 above)
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/payoutservice/exception/InternalException.kt:5]: This is used for Java's serialization mechanism. While we do not use it explicitly (and we shouldn't) it's a good practice to always define `serialVersionUID` for every class which implements `java.io.Serializable` - which is the case for exceptions since `Throwable` implements that interface.

**Review by githinho — APPROVED** [review]
